### ISSUE TITLE
MP: Viewing team of a client that controls two non-allied sides

### ISSUE THREAD

**Issue by jostephd** [wesnoth/wesnoth#4438]
Suppose there's a 3-way MP match and sides 2 and 3 are controlled by the same player, but are not allied. While it's side 1's turn, the player will have side 3's vision (and will speak in the chat with side 3's color).

It'll be better to give the player side 2's vision while it's side 1's turn, so he can plan his next moves. That is:

- [ ] As soon as the side turn of a remote (or [AI](https://github.com/wesnoth/wesnoth/issues/4330)) side starts, change wesnoth's view to the view of the _next_ side controller by the local player, rather than the previous one.

That should be relatively easy, and I'm labeling this as a good first issue. There is also another possible change, which is both more invasive and (I guess) more controversial:

- [ ] It might even make sense to let the player controlling sides 2 and 3 have the combined vision of side 2 and side 3 while it's not his turn. (However, it's probably neither possible nor practical to give him that vision during his own turns.)

**Comment by ProditorMagnus** [comment]
This is especially problematic when one of the sides is dead.

**Comment by xerus2000** [comment]
I was about to report this as an issue, but I guess it fits in here:
We played a three player match, I was faction 1. The leader of faction 2 died early and I was assigned control over it, so I selected it to use an AI, since I thought of it as a neutral force, not identifying myself with it.

Now, when it was the turn of faction 3, I had the vision of faction 2, which meant that
1) I saw things I wasn't supposed to see (vision deeper into faction 3 than my units allowed)
2) I DIDN'T SEE HOW MY UNITS DIED - i.e. when faction 3 attacked me (faction 1) in an area where faction 2 had no units, I did not see what was going on, only the result once it was my turn again. This can be my ruin!

**Comment by soliton-** [comment]
> This would indeed be rather easy to do, so is it agreed on that this would be a good change?

IMO yes.

**Comment by gfgtdf** [comment]
I think this could also simplify the code since currently it sometimes uses `first_human_team_` and sometimes `find_last_visible_team()`

**Comment by ProditorMagnus** [comment]
I think it should share vision between all sides that are local.

**Comment by gfgtdf** [comment]
> I think it should share vision between all sides that are local.

This would be a but harder to implement, do you have any use case for that one? I'd think that especially when the two local sides are enemies of each other "hot seat mp" this is not what you want, and if they are no, they probably already share vision.

**Comment by ProditorMagnus** [comment]
Yes, does not work in hot seat. But in regular games you can miss actions involving your unit no matter which side vision you use since one enemy side can attack both sides during one turn.

**Comment by gfgtdf** [comment]
What do you mean by "regular game" here?

**Comment by ProditorMagnus** [comment]
Game with at least one non-local side.

**Comment by gfgtdf** [comment]
Are these sides enemies it allies?

**Comment by ProditorMagnus** [comment]
Enemies -> remote player left, and because of that one client controls multiple enemy sides

**Comment by gfgtdf** [comment]
Enemies -> remote player left, and because of that one client controls multiple enemy sides

I still would think that this is a really strange (more uncommon than hotseat games at least) case: wouldn't you usually give control to a side that is allied to the side that left, and if there is no such side you'd basically be playing the game against yourselves (and aboodt the game)

**Comment by ProditorMagnus** [comment]
Consider 6 players, no teams. Once 2 are dead, 4 would still want to play. I have not checked recently how it behaves, but at some point dead sides still had to be assigned to player who would then lose vision. But it could be higher side or lower side number, which is why I think `first_human_team_ and sometimes find_last_visible_team()` is not solution.

**Comment by gfgtdf** [comment]
> but at some point dead sides still had to be assigned to player who would then lose vision

Can't you use "replace with ai" or "idle" for those? Afaik in both cases you don't loose vision

**Comment by gfgtdf** [comment]
But yes this is certainly a case that we should fix if it doesn't work

**Comment by gfgtdf** [comment]
Actually according to this report it currently might not work properly https://github.com/wesnoth/wesnoth/issues/1465

**Comment by ProditorMagnus** [comment]
Reproducing with 1.13.0 would not give much information though.

**Comment by CelticMinstrel** [comment]
I think if the player of a defeated side disconnects, you should do something like this.

1. If it has no units left, just delete it, set it to idle, whatever. It doesn't do anything and should no longer have any effect on the game.
2. Otherwise, if it's allied to the host, assign it to the host's control.
3. Otherwise (it's not allied to the host), assign it to control of the first player who's allied to it.
4. Otherwise (it's not allied to any human player), set it to be AI-controlled.

Or, alternatively, give the host a popup asking them which of these 3 things to do (only if it has units left though). Basically, the host has a choice of assigning it to a particular human player (with the list noting which ones it's allied to) or setting it to be AI-controlled.

**Comment by gfgtdf** [comment]
> Or, alternatively, give the host a popup asking them which of these 3 things to do (only if it has units left though). 

This is what currently happens (whether r defeated or not)

**Comment by gfgtdf** [comment]
> Reproducing with 1.13.0 would not give much information though.

The question is whether it's still an issue with 1.16 or not

**Comment by CelticMinstrel** [comment]
It looks like I jumped in halfway and misunderstood the issue. I think the use-case for this would actually be hotseat games? I mean, while normally a hotseat game is done locally, there's nothing preventing you from connecting to a server and inviting additional people, right? I think jostephd's proposal does kinda make sense.

**Comment by xeruf** [comment]
The issue I mentioned, my goodness is it old already, is unrelated to hotseat:
https://github.com/wesnoth/wesnoth/issues/4438#issuecomment-692279531

**Comment by gfgtdf** [comment]
> The issue I mentioned

It probably a duplicate of  #1465

**Comment by gfgtdf** [comment]
So the main part for this is now implement, for the behaviour when a player leaves/gives up  #1465 is still open

**Comment by BillCoder-cpu** [comment]
Why is this still open?  it works in the current release.  #1465 is related but remains open.

**Comment by gfgtdf** [comment]
Yes i agree
